src/_mssql.pyx

```diff
@@ -743,17 +743,29 @@ cdef class MSSQLConnection:
         with nogil:
             dbclose(self.dbproc)
 
-        self.mark_disconnected()
+        self._real_close()
 
-    def mark_disconnected(self):
-        log("_mssql.MSSQLConnection.mark_disconnected()")
-        self.dbproc = NULL
+    def _real_close(self):
+        """
+        Free resources used by this object.
+        """
+        log("_mssql.MSSQLConnection.real_close()")
+        if self in connection_object_list:
+            connection_object_list.remove(self)
         self._connected = 0
+        self.dbproc = NULL
         PyMem_Free(self.last_msg_proc)
         PyMem_Free(self.last_msg_srv)
         PyMem_Free(self.last_msg_str)
         PyMem_Free(self._charset)
-        connection_object_list.remove(self)
+
+    def mark_disconnected(self):
```

Comment:
Now this `mark_disconnected()` method is just a variable assignment I think we can remove it and replace its usage elsewhere (one occurrence) on this file with that assignment.